From d4601349716e80f945c3f3fc30211cbfb037c194 Mon Sep 17 00:00:00 2001
From: Murat Sezgin <msezgin@codeaurora.org>
Date: Mon, 21 Sep 2020 16:28:14 -0700
Subject: [PATCH] pppoe: Add return value to pppoe addressing get function

If the addressing doesn't have a netdevice, this should be
handled as a failure, so that the caller considers that
the get function failed.

Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
Change-Id: Ia9a6b0e0f036a3434519d9f2194763486ca04583
---
 drivers/net/ppp/pppoe.c  | 14 +++++++++-----
 include/linux/if_pppox.h |  4 ++--
 2 files changed, 11 insertions(+), 7 deletions(-)

--- a/drivers/net/ppp/pppoe.c
+++ b/drivers/net/ppp/pppoe.c
@@ -1037,24 +1037,28 @@ static int pppoe_get_channel_protocol(st
  * NOTE: This function returns a HOLD to the netdevice
  *
  ***********************************************************************/
-static void pppoe_get_addressing(struct ppp_channel *chan,
+static int pppoe_get_addressing(struct ppp_channel *chan,
 				 struct pppoe_opt *addressing)
 {
 	struct sock *sk = (struct sock *)chan->private;
 	struct pppox_sock *po = pppox_sk(sk);
+	int err = 0;
 
 	*addressing = po->proto.pppoe;
-	if (addressing->dev)
-		dev_hold(addressing->dev);
+	if (!addressing->dev)
+		return -ENODEV;
+
+	dev_hold(addressing->dev);
+	return err;
 }
 
 /* pppoe_channel_addressing_get()
  *	Return PPPoE channel specific addressing information.
  */
-void pppoe_channel_addressing_get(struct ppp_channel *chan,
+int pppoe_channel_addressing_get(struct ppp_channel *chan,
 				  struct pppoe_opt *addressing)
 {
-	pppoe_get_addressing(chan, addressing);
+	return pppoe_get_addressing(chan, addressing);
 }
 EXPORT_SYMBOL(pppoe_channel_addressing_get);
 
--- a/include/linux/if_pppox.h
+++ b/include/linux/if_pppox.h
@@ -101,7 +101,7 @@ enum {
 struct pppoe_channel_ops {
 	/* Must be first - general to all PPP channels */
 	struct ppp_channel_ops ops;
-	void (*get_addressing)(struct ppp_channel *, struct pppoe_opt *);
+	int (*get_addressing)(struct ppp_channel *, struct pppoe_opt *);
 };
 
 /* PPTP client callback */
@@ -109,7 +109,7 @@ typedef int (*pptp_gre_seq_offload_callb
 					       struct net_device *pptp_dev);
 
 /* Return PPPoE channel specific addressing information */
-extern void pppoe_channel_addressing_get(struct ppp_channel *chan,
+extern int pppoe_channel_addressing_get(struct ppp_channel *chan,
 					 struct pppoe_opt *addressing);
 
 /* Lookup PPTP session info and return PPTP session */
